ch5/ex-5-13-tail.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Exercise 5-13
+ *
+ * Write the program tail, which prints the last n lines of its input.  By
+ * default, n is 10, let us say, but it can be changed by an optional
+ * argument, so that
+ *
+ * 	tail -n
+ *
+ * prints the last n lines.  The program should behave rationally no matter
+ * how unreasonable the input or the value of n.  Write the program so it
+ * makes the best use of available storage; lines should be stored as in the
+ * sorting program of Section 5.6, not in a two-dimensional array of fixed
+ * size.
+ */
+
+#include <ctype.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define INITIAL_BUFFSIZE (1 << 10)
+#define DEFAULT_TAILSIZE 10
+
+enum { OK, ERROR };
+
+/* Single buffer to store characters, loops back to start on reaching limit */
+unsigned long buffsize;
+unsigned long maxbuffsize = ULONG_MAX;
+char *buff, *cursor, *bufflimit;
+
+/* max number of lines to print */
+unsigned long tailsize;
+/* pointers to the start of lines in the buffer, loops back start at limit */
+char **lines, **firstln, **currln, **lineslimit;
+
+int readlines(void);
+void writelines(void);
+int init_buff(unsigned long);
+int init_lines(void);
+int enlargebuff(void);
+int settailsize(int argc, char **argv);
+
+int main(int argc, char **argv)
+{
+	if (settailsize(argc, argv) != OK)
+		return 0;
```

Comment:
At this point in the book we're not explicitly returning values from main, I'm just doing it shut the compiler up.  But since I am making it explicit it wouldn't do any harm to make it meaningful.